libbeat/registry/backend/memlog/diskstore.go

```diff
@@ -0,0 +1,724 @@
+		s.logNeedsTruncate = true
+		return
+	}
+
+	err := s.logFile.Truncate(0)
+	if err == nil {
+		_, err = s.logFile.Seek(0, os.SEEK_SET)
+	}
+
+	if err != nil {
+		s.logFile.Close()
+		s.logFile = nil
+		s.logBuf = nil
+		s.logNeedsTruncate = true
+		s.logInvalid = true
+	}
+
+	s.logEntries = 0
+}
+
+// commitOps finishes a transaction by appending operations to the transaction
+// log file.
+// A transaction has a `Begin` and `Commit` operation, including a transaction
+// IDs. On error a `Rollback` operation will be inserted.
+//
+// If an error was encountered, that will leave the transaction log in a
+// potential broken state, we force a checkpoint based commit within the next
+// transaction. The current transaction will be marked as failed.
+//
+// The txid will be incremented upon success.
+func (s *diskStore) commitOps(st *txState) error {
+	writer := s.logBuf
+	enc := json.NewEncoder(writer)
+
+	ok := false
+	defer func() {
+		if !ok {
+			// writing to log file failed -> fail current transaction and force full
+			// checkpoint on next transaction
+			s.logInvalid = true
+		}
+	}()
+
+	err := encOp(enc, &opBegin{ID: s.txid + 1})
+	if err != nil {
+		return err
+	}
+
+	var total uint
+	for _, bin := range st.bins {
+		for i := range bin {
+			entry := &bin[i]
+			if !entry.modified {
+				continue
+			}
+
+			for _, op := range entry.ops {
+				if err = encOp(enc, op); err != nil {
+					return nil
+				}
+			}
+
+			total += uint(len(entry.ops))
+		}
+	}
+
+	err = encOp(enc, &opCommit{ID: s.txid + 1})
+	if err != nil {
+		return err
+	}
+
+	err = writer.Flush()
+	if err != nil {
+		return err
+	}
+
+	ok = true
+	s.txid++
+	s.logEntries += total
+	return nil
+}
+
+func newDiskLoader() *diskLoader {
+	return &diskLoader{
+		typeConv: newTypeConv(),
+	}
+}
+
+func (l *diskLoader) IsReadonly() bool       { return false }
+func (l *diskLoader) checkRead() error       { return nil }
+func (l *diskLoader) checkWrite() error      { return nil }
+func (l *diskLoader) getTypeConv() *typeConv { return l.typeConv }
+
+// loadDataFile create a new hashtable with all key/value pairs found.
+func loadDataFile(path string) (hashtable, error) {
+	var tbl hashtable
+	tbl.init()
+
+	if path == "" {
+		return tbl, nil
+	}
+
+	hashFn := newHashFn()
+	err := readDataFile(path, func(key []byte, state common.MapStr) {
+		tbl.set(hashFn(key), key, state)
+	})
+	return tbl, err
+}
+
+// loadLogFile applies all recorded transaction to an already initialized
+// memStore.
+// The txid is the transaction ID of the last known valid data file.
+// Transactions older then txid will be ignored.
+// loadLogFile returns the last commited txid in logTxid and the total number
+// of operations in logCount.
+// An incomplete transaction is recorded at the end of the log file, if
+// complete is false.
+func loadLogFile(
+	store *memStore,
+	txid uint64,
+	home string,
+) (logTxid uint64, logCount uint, complete bool, err error) {
+	var tx *memTx
+	var entries uint
```

Comment:
I was tracking down the **entries** vars, _opInsertWith_, _opUpdate_, _opRemove_ is incrementing this value. The value is used https://github.com/elastic/beats/pull/8131/files#diff-a58812485a14d0865d2e0cd3753f765dR96 when we create the new disk store. There is an accessor at https://github.com/elastic/beats/pull/8131/files#diff-a58812485a14d0865d2e0cd3753f765dR165

The purpose of entries is to track the number of operation in the oplog?

Reply:
Right. The entry count is passed to the 'checkpoint' check to decide when to do a checkpoint. Will create a stats structure in future PRs, so we can also take file sizes into account.